ratik/src/main/java/org/opencadc/ratik/InventoryValidator.java

```diff
@@ -150,7 +169,215 @@ public InventoryValidator(Map<String, Object> daoConfig, URI resourceID, Artifac
     }
 
     @Override public void run() {
-        throw new UnsupportedOperationException("NOT IMPLEMENTED");
+        try {
+            final Subject subject = SSLUtil.createSubject(new File(CERTIFICATE_FILE_LOCATION));
+            Subject.doAs(subject, (PrivilegedExceptionAction<Void>) () -> {
+                doit();
+                return null;
+            });
+        } catch (PrivilegedActionException privilegedActionException) {
+            final Exception exception = privilegedActionException.getException();
+            throw new IllegalStateException(exception.getMessage(), exception);
+        }
+    }
+
+    /**
+     * Validates local and remote sets of Artifacts.
+     *
+     * @throws ResourceNotFoundException For any missing required configuration that is missing.
+     * @throws IOException               For unreadable configuration files.
+     * @throws IllegalStateException     For any invalid configuration.
+     * @throws TransientException        temporary failure of TAP service: same call could work in future
+     * @throws InterruptedException      thread interrupted
+     */
+    void doit() throws ResourceNotFoundException, IOException, IllegalStateException, TransientException,
+                       InterruptedException {
+
+        Iterator<String> bucketIterator = this.bucketSelector.getBucketIterator();
+        while (bucketIterator.hasNext()) {
+            final String bucket = bucketIterator.next();
+            log.debug("processing bucket: " + bucket);
+            try (final ResourceIterator<Artifact> localIterator = getLocalIterator(bucket);
+                final ResourceIterator<Artifact> remoteIterator = getRemoteIterator(bucket)) {
+
+                Artifact localArtifact = null;
+                Artifact remoteArtifact = null;
+                boolean artifactsToValidate = true;
+                while (artifactsToValidate) {
+                    if (localArtifact == null) {
+                        localArtifact = localIterator.hasNext() ? localIterator.next() : null;
+                    }
+                    if (remoteArtifact == null) {
+                        remoteArtifact = remoteIterator.hasNext() ? remoteIterator.next() : null;
+                    }
+                    // TODO sanity check? if either iterator has no results in the first loop, exit?
+                    if (localArtifact == null && remoteArtifact == null) {
+                        artifactsToValidate = false;
+                        continue;
+                    }
+                    log.debug(String.format("comparing Artifacts:\n local - %s\nremote - %s",
+                                            localArtifact, remoteArtifact));
+
+                    // check if Artifacts are the same, or if the local Artifact
+                    // precedes or follows the remote Artifact.
+                    int order = orderArtifacts(localArtifact, remoteArtifact);
```

Comment:
I would use a switch statement here: explicitly -1, 0, 1, and then `default` being a failure

Reply:
The orderArtifacts(local, remote) method now returns -1,0,1 instead of <neg int>,0,<pos int>, for use with a switch statement.